### PR TITLE
Contextual helpers

### PR DESCRIPTION
**Pull request by cibernox**
[rendered](https://github.com/cibernox/rfcs/blob/contextual-helpers/text/0000-contextual-helpers.md)

### CODE REVIEW

File: text/0000-contextual-helpers.md [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+Examples:
+
+```hbs
+<strong>My name is {{invoke-helper 'capitalize-str' user.name}}</strong>
+<button title={{invoke-helper 'capitalize-str' user.name}}></button>
+<strong>My name is {{helper 'capitalize-str' user.name}}</strong> <!-- Invalid invocation -->
+<button title={{helper 'capitalize-str' user.name}}></button> <!-- Invalid invocation -->
+
+{{my-component format-date=(helper "moment-format")}} <!-- Creates closure helper -->
+{{my-component format-date=(invoke-helper "moment-format")}} <!-- Invokes helper -->
```
Comment (NLincoln) [inline text/0000-contextual-helpers.md]: What happens to the existing syntax for invoking helpers, which for this example would be:

```hbs
{{my-component format-date=(moment-format)}}
```

Comment (cibernox) [inline text/0000-contextual-helpers.md]: that would invoke the helper which generates a string that is passed to the component, like today.

**Comment by knownasilya** [comment]
Yes please, and then deprecate `{{my-helper}}` so we don't have that ambiguity.. is it a value, component, or helper?

**Comment by miguelcobain** [comment]
It can be any of those, but there are priorities in its resolution. value, helper and then component, I think.

**Comment by knownasilya** [comment]
Yes, but from a developer point of view, it's ambiguous. Especially for developers new to Ember.

**Comment by miguelcobain** [comment]
A bit unrelated to the RFC, I guess, but I'm not sure how to solve that.

**Comment by machty** [comment]
There's an inconsistency in helper lifespans that I've noticed in a few other cases that we might want to consider as part of this RFC: https://ember-twiddle.com/d869daef1f0c2c8e2808a8e60dc06ca3?numColumns=2&openFiles=templates.application.hbs%2Chelpers.count-up.js

**Comment by ming-codes** [comment]
An helper is mostly just a function. Can the problem be solved by simply yielding a function (or action)?

```hbs
{{yield
  (action "capitalize-str")
}}
```

```hbs
{{#parent-component as |capitalize-str|}}
  {{child-component helper=capitalize-str}}
{{/parent-component}}
```

Only down side I see is that you can't create a helper that maintain it's own state.

**Comment by cibernox** [comment]
@ming-codes It could, and [it's what I'm doing for now](https://github.com/cibernox/ember-power-datepicker/blob/master/addon/templates/components/power-datepicker.hbs#L39), but there is no built in way of invoking that helper in the template, so I have to use ember-composable-helper's [compute](https://github.com/DockYard/ember-composable-helpers#compute) utility.

This proves that it's technically possible, but:

- Syntax is 👎 
- It doesn't support stateful helpers
- It doesn't support, like in contextual components, to override passed default values

Hence, I think there is room for a feature here.

**Comment by knownasilya** [comment]
Does https://github.com/emberjs/rfcs/pull/432 supersede this RFC?

**Comment by cibernox** [comment]
@knownasilya maybe? I haven't seen that until now.

**Comment by chancancode** [comment]
As others have pointed out, I think the design and exploration here have been absorbed into #432. Let's move the discussion there. Thanks again @cibernox for pioneering the work in this area!

**Pull request closed without merging**